src/spatial/utils.jl

```diff
@@ -26,191 +22,358 @@ abstract type AbstractSpatialSystem end
+    @inbounds u[species,site]*hop_rates.hopping_constants[species,site]*num_nbs
+end
+
+############## hopping rates of form L_{s,i,j} ################
+struct HopRatesGeneral{F} <: AbstractHopRates
+    hopping_constants::Vector{Matrix{F}} # hopping_constants[i][j,s] is the hopping constant at site i of species s to jth neighbor
+    rates::Vector{Matrix{F}} # rates[i][j,s] is the hopping rate at site i of species s to jth neighbor
+    sum_rates::Vector{F} # sum_rates[i] is the total hopping rate out of site i
 end
 
 """
-initializes SpatialRates with zero rates
+initializes HopRates with zero rates
 """
-function SpatialRates(ma_jumps, num_species, spatial_system::AbstractSpatialSystem)
-    num_sites = num_sites(spatial_system)
-    num_jumps = get_num_majumps(ma_jumps)
-    SpatialRates(num_jumps,num_species,num_sites)
+function HopRatesGeneral(hopping_constants::Vector{Matrix{F}}, transpose = true) where F <: Number
+    if transpose
+        hopping_constants = map(r -> r = collect(r'), hopping_constants) #transpose for better memory layout
+    end
+    rates = deepcopy(hopping_constants)
+    foreach(r -> r .= zero(F), rates)
+    sum_rates = zeros(F, length(rates))
+    HopRatesGeneral{F}(hopping_constants, rates, sum_rates)
 end
 
 """
 make all rates zero
 """
-function reset!(spatial_rates)
-    fill!(spatial_rates.rx_rates, zero(eltype(spatial_rates.rx_rates)))
-    fill!(spatial_rates.hop_rates, zero(eltype(spatial_rates.hop_rates)))
-    fill!(spatial_rates.rx_rates_sum, zero(eltype(spatial_rates.rx_rates_sum)))
-    fill!(spatial_rates.hop_rates_sum, zero(eltype(spatial_rates.hop_rates_sum)))
+function reset!(hop_rates::HopRatesGeneral{F}) where F <: Number
+    foreach(r -> fill!(r, zero(F)), hop_rates.rates)
+    hop_rates.sum_rates .= zero(F)
     nothing
 end
+
 """
-return total reaction rate at site
+sample a reaction at site, return (species, target_site)
 """
-function total_site_rx_rate(spatial_rates, site_id)
-    spatial_rates.rx_rates_sum[site_id]
+function sample_hop_at_site(hop_rates::HopRatesGeneral, site, rng, spatial_system) 
+    @inbounds rates_at_site = hop_rates.rates[site]
+    r = rand(rng) * total_site_hop_rate(hop_rates, site)
+    n, species = Tuple(CartesianIndices(rates_at_site)[linear_search(rates_at_site, r)])
+    return species, nth_nbr(spatial_system, site, n)
 end
 
 """
-return total hopping rate out of site
+update rates of single species at site
 """
-function total_site_hop_rate(spatial_rates, site_id)
-    spatial_rates.hop_rates_sum[site_id]
+function update_hop_rate!(hop_rates::HopRatesGeneral, species, u, site, spatial_system)
+    rates_at_site = hop_rates.rates[site]
+    @inbounds old_rate = sum(@view rates_at_site[:,species])
+    @inbounds @. rates_at_site[:,species] = (@view hop_rates.hopping_constants[site][:,species]) * u[species,site]
+    @inbounds hop_rates.sum_rates[site] += sum(@view rates_at_site[:,species]) - old_rate
```

Comment:
Maybe a better way to optimize this is to store the partial sums of the hopping constants instead of the constants themselves, and then use those for sampling the next direction? Then we just multiply their already stored total sum, which is fixed, by the new amount of the species to update the `sum_rate`? Wouldn't that avoid all these vector operations?